### DISCUSSION TITLE
org-ref and citar/org-cite compatibility and features

### DISCUSSION THREAD

**Discussion by deb75** [emacs-citar/citar#725] [Q&A]
Hi,

I would like to jump from `org-ref` to `citar/org-cite`.

I am just wondering if the citations in my org files, which are based on `org-ref` : [[cite:&citationkey]]
will always be valid with `citar/org-cite` ?

Furthermore, here is how I configure `org-ref` for a particular org file
```
#+LATEX_HEADER: \usepackage[citestyle=authoryear-icomp,bibstyle=authoryear, hyperref=true,backref=true,maxcitenames=3,url=false,backend=biber,natbib=true]{biblatex}
#+LATEX_HEADER: \addbibresource{d:/edebry/Documents/reference/bibliography.bib}
```
Do I need to change this for `citar/org-cite` ?

Also, a nice feature of `org-ref` I appreciate is the handling of references, especially that of math
environnments. Does `citar/org-ref` manages these references ?

Regards

**Comment by bdarcus** [comment]
> I would like to jump from `org-ref` to `citar/org-cite`.

Cool!

> I am just wondering if the citations in my org files, which are based on `org-ref` : [[cite:&citationkey]] will always be valid with `citar/org-cite` ?

You may want to read through this page to get a general sense of the place of this package.

https://github.com/emacs-citar/citar/wiki/Comparisons#citar-in-context

On your question, citar  defers to org itself for what is and isn't a citation (org does the parsing etc). From that POV, it isn't.

But you can also configure custom major mode functions, though that only covers insertion (not at-point actions):

https://github.com/emacs-citar/citar/wiki/major-mode-functions
 
> Furthermore, here is how I configure `org-ref` for a particular org file
> 
> ```
> #+LATEX_HEADER: \usepackage[citestyle=authoryear-icomp,bibstyle=authoryear, hyperref=true,backref=true,maxcitenames=3,url=false,backend=biber,natbib=true]{biblatex}
> #+LATEX_HEADER: \addbibresource{d:/edebry/Documents/reference/bibliography.bib}
> ```
> 
> Do I need to change this for `citar/org-cite` ?

I don't have time to look into the details on this ATM, but maybe check the org manual? There's also a few good blog posts on `org-cite` that should explain this. It's easy.

It does sound like you want to use the `oc-biblatex` export processor, however (if you're using biblatex).

> Also, a nice feature of `org-ref` I appreciate is the handling of references, especially that of math environnments. Does `citar/org-ref` manages these references ?

You mean cross-references?

No; I consider that out of scope; see #427.

PS - I may convert this to a discussion.

**Reply by bdarcus** [reply]
A very recent tutorial, but using `oc-csl`.

https://andras-simonyi.github.io/org-csl-cv-bib-tutorial/

**Comment by notuntoward** [comment]
I'd also like to switch to `org-cite`, but I haven't done it yet because the way its links are displayed in emacs will mess up a giant .org file I've kept going for almost a decade.  Or am I wrong about that?

In emacs, an `org-ref` link written like this:  

> ... an alternative is `[[cite:Ying22multiTaskTreeMTGB][Ying et al. 2022: MT-GBM multitask tree]]`, which solves...

can display in an emacs buffer something like this:  

> ... an alternative is _Ying et al. 2022: MT-GBM multitask tree_, which solves...

I remember when `org-cite` was first proposed, that @bdarcus thought that this would be doable in `org-cite`.

But I can't figure out how to do it.  Can it be done?

**Reply by bdarcus** [reply]
Yes.

That's the domain for `org-cite` "activate" processors. See, for example:

https://github.com/andras-simonyi/org-cite-csl-activate

**Reply by notuntoward** [reply]
Thanks!  But it looks like development on `org-cite-csl-activate` has stopped, with significant problems remaining.  Google didn't reveal other packages like this.  Is this the only one you know of?

**Reply by bdarcus** [reply]
Not AFAIK. Hoping people will start providing them, as org-cite is a wonderfully modular system easy to develop for.

@andras-simonyi - any input? I don't recall the status of your processor.

**Reply by andras-simonyi** [reply]
Development has stopped in the sense that I've solved all problems which seemed important and reasonably solvable, and the package has been pretty stable with no bugs reported for a while. I am actively using it, and there seems to be a small group of people who are also active users. Actually, I've been planning to add it to MELPA , just haven't had the time yet to do so.

As for the remaining problems, none of those I know about are show-stoppers for me, but can be, of course, for others; it's difficult to know without any submitted issues.  It is  true that the processor in its current form is not really usable for Org files with a very large number of citations (say more than 500 or 1000) if all citations are to be rendered; this use case probably requires a different way of generating previews. (Either rendering only locally or using a cheaper rendering method.)

> ... an alternative is [[cite:Ying22multiTaskTreeMTGB][Ying et al. 2022: MT-GBM multitask tree]], which solves...

where does the citation link's text come from? Is it entered by the user, or automatically generated?

**Reply by notuntoward** [reply]
The text is first automatically generated by org-ref.  I think it's using the same method as the old bibtex links use: 

> Author year: chunks of the title

Then I sometimes edit the description to something that says more specifically why I'm linking to this reference.  Kind of like how journalists write link descriptions on the web.

Here's how I have the autogenerated text configured:
```
  ;; bibtex-key generator: firstauthor-year-title-words
  (setq bibtex-autokey-year-length 4
        bibtex-autokey-name-year-separator "-"
        bibtex-autokey-year-title-separator "-"
        bibtex-autokey-titleword-separator "-"
        bibtex-autokey-titlewords 2
        bibtex-autokey-titlewords-stretch 1
        bibtex-autokey-titleword-length 5)
```

**Reply by bdarcus** [reply]
> Then I sometimes edit the description to something that says more specifically why I'm linking to this reference. Kind of like how journalists write link descriptions on the web.

You can't do annotations like that in a citation, unless it's affix text destined for output; e.g.:

```
Foo bar [cite:for more on X, see @doe].
```

**Reply by notuntoward** [reply]
Thanks for that info.  Is it true, then, that in this case, `org-cite-csl-activatate` would display in emacs as

>  _for more on x, see_

or would it be

>  _for more on x, see Doe 2022_

supposing that Doe wrote something in 2022?

**Reply by bdarcus** [reply]
The latter, with the precise details depending on [what CSL style you use for the preview](https://github.com/andras-simonyi/org-cite-csl-activate#csl-style-and-locale).

**Reply by notuntoward** [reply]
That would cover maybe 80% of the links in my big org file of references.  Is this the place to request a feature that would implement the org-ref or old bibtex mode behavior?  Sometimes, for flow and minimal distraction, arbitrary text is just better than link text with an author and year in it.

**Reply by andras-simonyi** [reply]
Org-cite's citation suffixes and prefixes can contain arbitrary text and are also displayed as part of the rendered citations. Of course, arbitrary text is more flexible, but an important point to think about is exporting the document: `org-cite-csl-activatate` shows citations as they would be rendered in exported documents (using the CSL export processor) -- I suspect org-ref just ignores the citation link texts during export.

**Reply by bdarcus** [reply]
Keep in mind too that the tooltip in `org-cite-csl-activate` displays the full bibliographic entry.

A citation reference has affixes (that as @andras-simonyi says are exported) and the key, plus an optional style, and that's it. I don't see how a feature request (to org really) to add per-reference annotations would even work technically. What you seem to be suggesting is a completely different model than the current design. 

Maybe you should just experiment with `org-cite-csl-activate`, @notuntoward, without initially worrying about converting  existing documents? 

The beauty of `org-cite` to me is its modular nature. You can easily change the activate processor to get different behavior there, and even write your own if you want, without `citar` or the export processors having to know anything about that.

> ... old bibtex mode behavior.

What's this though?

**Comment by notuntoward** [comment]
_"Old bibtex mode behavior"_:  I'm talking about the behavior I get if I don't install org-ref or citar: what happens when I put the cursor on a bibtex file entry and run `M-x org-store-link`, then paste the link into an org fiile with  'M-x org-insert-link'  The format of the resulting org link is like:

> [[file:energy.bib::Ying22multiTaskTreeMTGB][Ying et al. 2022: MT GBM]]

I guess that's just an ordinary file link but with the auto-generated bibtex-specific link text.  I used these for years because how the link was rendered in org was totally editable.  Eventually, I switched to org-ref for all the perks.  There, it was also possible to get arbitrary link text rendering, at the cost of complicating exports.  But complicated exports was not a problem because...

Exporting a technical, referenced article from org-mode is something I almost never do.  I tried to make this work, but ended up doing my PhD and first papers in plain latex, because that's what my advisor wanted.  Then in industry, anything published with references had to be in Word.  Then back in academia, my students and colleagues all over the place universally wanted to use online latex editors, like [Overleaf](https://www.overleaf.com/).  Org-mode just wasn't feasible.

But I love org-mode.  It's where I keep almost all of my brainstorming and literature notes.  Org is just for me.  For that use case, it's best to have arbitrary link text displayed in org-mode, as org-mode is the only way I ever see my notes.

Still, the new emacs cite links have even more perks than org-ref.  I'd love to get that working.

**Reply by bdarcus** [reply]
In that example, what is `MT GBM`?

If you don't care about export, then it really doesn't matter what the answer is; you could likely do this:

```
[cite:@Ying22multiTaskTreeMTGB MT GBM]
```

Then `org-cite-csl-activate` will render that basically the same, depending on which CSL style you choose for rendering.

Affixes like that are usually used for things like page numbers, that of course do export, and whose rendering will vary by citation style. But since you don't care, you can use them for any kind of citation/citation-reference-specific annotation.
